added Tuscan square and alphabetical test ordering

mmk-1 opened the pull request:
Hello,
I have added TuscanDetector.java and AlphabeticalDetector.java as the new detector types for iDFlakies. 
The alphabetical detector will order the test classes based on their alphabetical order.
The tuscan square detector will create the tuscan permutations of test classes and uses each permutation each round. The number of rounds is the minimum of the number of full permutations or the given number of rounds.
For Tuscan square ordering: 
`-Ddetector.detector_type=tuscan`
For alphabetical ordering: 
`-Ddetector.detector_type=alphabetical`

august782 commented on idflakies-core/src/main/java/edu/illinois/cs/dt/tools/detection/TestShuffler.java:
```diff
@@ -194,4 +195,39 @@ private int permutations(final int accum, final Iterator<List<String>> iterator,
+        int[][] res = t.getMatrix();
+        List<String> fullTestOrder = new ArrayList<String>();
+        List<String> permClasses = new ArrayList<String>();
+        if(n == 3 || n == 5) {
```
Space after `if`

august782 commented on idflakies-core/src/main/java/edu/illinois/cs/dt/tools/detection/TestShuffler.java:
```diff
@@ -194,4 +195,39 @@ private int permutations(final int accum, final Iterator<List<String>> iterator,
+                permClasses.add(classes.get(res[count][i]));
+            }
+        } else {
+            int nn = res.length-1;
```
Space around `-`

august782 commented on idflakies-core/src/main/java/edu/illinois/cs/dt/tools/detection/detectors/AlphabeticalDetector.java:
```diff
@@ -0,0 +1,43 @@
+public class AlphabeticalDetector extends ExecutingDetector {
+    final List<String> tests;
+    private TestRunResult origResult;
+    private TestRunResult lastAlphabeticalResult;
```
I don't think you actually needs these last results for AlphabeticalDetector. I believe these last results were used to keep track of what random order was tried previously, and since this detector only has one valid order, it doesn't really need to track that.

august782 commented on idflakies-core/src/main/java/edu/illinois/cs/dt/tools/detection/detectors/DetectorFactory.java:
```diff
@@ -24,8 +36,16 @@ public static Detector makeDetector(final InstrumentingSmartRunner runner, final
             return new SmartShuffleDetector(runner, baseDir, rounds, tests, detectorType());
+        }  else if (detectorType().equals("tuscan")){
+            int n = getClassesSize(tests);
+            if(n == 3 || n == 5) {
```
Space after `if`

august782 commented on idflakies-core/src/main/java/edu/illinois/cs/dt/tools/detection/detectors/DetectorFactory.java:
```diff
@@ -24,8 +36,16 @@ public static Detector makeDetector(final InstrumentingSmartRunner runner, final
+            return new AlphabeticalDetector(runner, baseDir, rounds, detectorType(), tests);
         }
-
+        
```
Don't need blank line

august782 commented on idflakies-core/src/main/java/edu/illinois/cs/dt/tools/detection/detectors/DetectorFactory.java:
```diff
@@ -24,8 +36,16 @@ public static Detector makeDetector(final InstrumentingSmartRunner runner, final
+        }  else if (detectorType().equals("tuscan")){
+            int n = getClassesSize(tests);
+            if(n == 3 || n == 5) {
+                return new TuscanDetector(runner, baseDir, rounds > (n+1) ? (n+1) : rounds, detectorType(), tests);
```
Does it have to distinguish number of rounds here upon creation? Can it be handled within the detector itself (so the additional code to be modified in this existing class can be simplified)?

mmk-1 replied:
I have moved them into TuscanDetector.java and it worked same as before.

august782 commented on idflakies-core/src/main/java/edu/illinois/cs/dt/tools/utility/Tuscan.java:
```diff
@@ -0,0 +1,168 @@
+import scala.annotation.meta.getter;
+
+public class Tuscan {
+    public void main(int arg) {
```
Maybe change this `main` method name to something more descriptive, like `generateTuscanSquare`.

august782 commented on idflakies-core/src/main/java/edu/illinois/cs/dt/tools/utility/Tuscan.java:
```diff
@@ -0,0 +1,168 @@
+    }
+    static void generateTuscanSquare(int n) {
+    	int nn = n;
+    	while((n-1) % 4 == 0 && n != 1 && n != 9) n = (n-1)/2+1;
```
Spaces around operators.

august782 commented on idflakies-core/src/main/java/edu/illinois/cs/dt/tools/utility/Tuscan.java:
```diff
@@ -0,0 +1,168 @@
+        }
+    }
+    static void generateThree() {
+        // int[][] t = {
```
Is this here so we can decide whether we want to have permutations that consist of all tests, or we want permutations with just the subset of tests such that all pairs are covered?

mmk-1 replied:
Yes I wasn't sure which one would be the appropriate permutation to use, so I thought to keep the permutation with subset of tests as a commented part just in case.

august782 replied:
Even though it's a comment, can just add a 0 to the end of each row just to match the uncommented part, just in case we change.

august782 commented on idflakies-core/src/main/java/edu/illinois/cs/dt/tools/utility/Tuscan.java:
```diff
@@ -0,0 +1,168 @@
+                { 0, 1, 2 },
+                { 1, 0, 2 },
+                { 2, 0, 1 },
+                { 2, 1, 0}
```
Space after 0, for consistency.

august782 commented on idflakies-core/src/main/java/edu/illinois/cs/dt/tools/utility/Tuscan.java:
```diff
@@ -0,0 +1,168 @@
+            }
+            helper(t, n-1);
+        } else if (n == 9) {
+            int[][] t = {{0,1,7,2,6,3,5,4,8},
```
Technically, this is inconsistent with the later hard-coded matrices since this one has no spaces, but I don't know how good it looks if you were to add spaces here...

august782 commented on idflakies-core/src/main/java/edu/illinois/cs/dt/tools/utility/Tuscan.java:
```diff
@@ -0,0 +1,168 @@
+            for (int i = 1; i < n; i++) {
+                t[i] = b[t[i-1]];
+            }
+            helper(t, n-1);
```
Spaces around operator.

august782 commented on idflakies-core/src/main/java/edu/illinois/cs/dt/tools/utility/Tuscan.java:
```diff
@@ -0,0 +1,168 @@
+                for (int j = h; j < n + 1; j++) {
+                    r[i][j] = ((j % 2 == 0) ? 0 : h) + r[i][j-h] % h;
+                }
+                // System.out.println(java.util.Arrays.toString(r[i]));
```
Can remove the print.

august782 commented on idflakies-core/src/main/java/edu/illinois/cs/dt/tools/utility/Tuscan.java:
```diff
@@ -0,0 +1,168 @@
+        } else {
+            generateTuscanSquare(arg);
+        }
+    }
```
Empty lines between methods and fields.

august782 commented on idflakies-core/src/main/java/edu/illinois/cs/dt/tools/utility/Tuscan.java:
```diff
@@ -0,0 +1,168 @@
+                { 2, 1, 0}
+        };
+
+        r = new int[4][];
```
Why not just let `r = t;`?

august782 commented on idflakies-core/src/main/java/edu/illinois/cs/dt/tools/utility/Tuscan.java:
```diff
@@ -0,0 +1,168 @@
+                { 0, 4, 1, 3, 2},
+                { 4, 0, 3, 1, 2}
+        };
+        r = new int[6][];
```
Why not just let `r = t;`?

mmk-1 commented:
Hello, I have fixed the issues you mentioned. Can you please check?
Thank you

mmk-1 commented on idflakies-core/src/main/java/edu/illinois/cs/dt/tools/detection/detectors/AlphabeticalDetector.java:
```diff
@@ -0,0 +1,40 @@
+import edu.illinois.cs.testrunner.runner.Runner;
+
+public class AlphabeticalDetector extends ExecutingDetector {
+    final List<String> tests;
```
private

mmk-1 commented on idflakies-core/src/main/java/edu/illinois/cs/dt/tools/detection/detectors/AlphabeticalDetector.java:
```diff
@@ -0,0 +1,40 @@
+    public DetectionRound results() throws Exception {
+        return makeDts(origResult, runList(testShuffler.alphabeticalOrder()));
+    }
+    
```
remove line

mmk-1 commented on idflakies-core/src/main/java/edu/illinois/cs/dt/tools/detection/detectors/TuscanDetector.java:
```diff
@@ -0,0 +1,62 @@
+import edu.illinois.cs.testrunner.runner.Runner;
+
+public class TuscanDetector extends ExecutingDetector {
+    final List<String> tests;
```
private

mmk-1 commented on idflakies-core/src/main/java/edu/illinois/cs/dt/tools/detection/detectors/TuscanDetector.java:
```diff
@@ -0,0 +1,62 @@
+        super(runner, baseDir, rounds, type);
+        int n = getClassesSize(tests);
+        if (n == 3 || n == 5) {
+            if(this.rounds > n) this.rounds = n + 1;
```
make if into code block

mmk-1 replied:
and space after if

mmk-1 commented on idflakies-core/src/main/java/edu/illinois/cs/dt/tools/detection/detectors/TuscanDetector.java:
```diff
@@ -0,0 +1,62 @@
+    public DetectionRound results() throws Exception {
+        lastTuscanResult = runList(testShuffler.tuscanOrder(absoluteRound.get()));
+        lastTuscanDetectionRound = makeDts(origResult, lastTuscanResult);
+        return lastTuscanDetectionRound;
```
make results() into one line

mmk-1 commented:
Hello, I have addressed the issues discussed in the new commit. The only thing is that in `TestShuffler.java` line 207 I need to make the distinction between n=3 or n=5 and the else case, because in the `Tuscan.java` `generateTuscanSquare()` which was implemented before, each row is instantiated into n+1 spaces therefore there is always an extra 0 as the last element. for example for n=2, the array will look like this: `{{1,2,0}, {2,1,0}}`
With my inspections it is because of line 27 inside Tuscan.java. However, I did not modify `generateTuscanSquare()` and rather made this if-else block as I feared it might break the algorithm to construct Tuscan squares.

august782 commented on idflakies-core/src/main/java/edu/illinois/cs/dt/tools/utility/Tuscan.java:
```diff
@@ -0,0 +1,149 @@
+    }
+
+    private static void generateThree() {
+        int[][] t = {
```
I wonder if you could just add an extra 0 to each row, just so it matches how the actual method has an extra 0 at the end of each row. Then some of the logic in other classes can be simplified instead of constantly checking whether size is 3 or 5.

august782 commented on idflakies-core/src/main/java/edu/illinois/cs/dt/tools/utility/Tuscan.java:
```diff
@@ -0,0 +1,149 @@
+                int[] t = new int[n];
+                System.arraycopy(r[i], l+1, t, 0, n - l);
+                System.arraycopy(r[i], 0, t, n - l, l);
+		        System.arraycopy(t, 0, r[i], 0, n);
```
Some mix of tabs and spaces here, use spaces consistently.

august782 commented on idflakies-core/src/main/java/edu/illinois/cs/dt/tools/utility/Tuscan.java:
```diff
@@ -0,0 +1,149 @@
+    private static int[][] r;
+
+    private static void helper(int[] a, int i) {
+	    System.arraycopy(a, 0, r[i], 0, a.length);
```
Uses spaces instead of tabs.

august782 commented on idflakies-core/src/main/java/edu/illinois/cs/dt/tools/utility/Tuscan.java:
```diff
@@ -0,0 +1,149 @@
+    	while((n - 1) % 4 == 0 && n != 1 && n != 9){
+             n = (n - 1) / 2 + 1;
+        }
+	    r = new int[nn][];
```
Use spaces instead of tabs.

august782 commented on idflakies-core/src/main/java/edu/illinois/cs/dt/tools/utility/Tuscan.java:
```diff
@@ -0,0 +1,149 @@
+            }
+            helper(t, n - 1);
+        } else if (n == 9) {
+            int[][] t = { 
```
Are we sure that in the case of n == 9 that it will also ensure each row has that extra 0 at the end?

august782 commented on idflakies-core/src/main/java/edu/illinois/cs/dt/tools/detection/TestShuffler.java:
```diff
@@ -194,4 +195,32 @@ private int permutations(final int accum, final Iterator<List<String>> iterator,
+            int n = classes.size();
+            int[][] res = Tuscan.generateTuscanPermutations(n);
+            List<String> permClasses = new ArrayList<String>();
+            if (n == 3 || n == 5) {
```
If you end up simplifying and always have it return a matrix where the last column of each row is 0, then can simplify this conditional.

august782 commented on idflakies-core/src/main/java/edu/illinois/cs/dt/tools/detection/detectors/TuscanOnlyClassDetector.java:
```diff
@@ -0,0 +1,61 @@
+    public TuscanOnlyClassDetector(final Runner runner, final File baseDir, final int rounds, final String type, final List<String> tests) {
+        super(runner, baseDir, rounds, type);
+        int n = getClassesSize(tests);
+        if (n == 3 || n == 5) {
```
Leave some comment here as a reminder that we need one more round than the number of test classes if n is 3 or 5.

august782 commented on idflakies-core/src/main/java/edu/illinois/cs/dt/tools/detection/detectors/TuscanOnlyClassDetector.java:
```diff
@@ -0,0 +1,62 @@
+        if (n == 3 || n == 5) {
+            // We need one more round than the number of classes if n is 3 or 5.
+            if (this.rounds > n) {
+                    this.rounds = n + 1;
```
Too much indentation

winglam commented:
Looks good to August and me. Thanks a LOT for the work!
